File: src/test/java/com/letgo/book/acceptance/TestCase.java
package com.letgo.book.acceptance;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.http.*;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public abstract class TestCase {
    public static final String DOMAIN_NAME = "http://localhost:";
    @LocalServerPort
    protected int port;
    @Autowired
    protected TestRestTemplate restTemplate;

    protected ResponseEntity<String> get(String endpoint) {
        return restTemplate.getForEntity(DOMAIN_NAME + port + endpoint, String.class);
    }

    protected ResponseEntity<String> post(String body) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        HttpEntity<String> entity = new HttpEntity<>(body, headers);

        return restTemplate.postForEntity(DOMAIN_NAME + port + "/book", entity, String.class);
    }

    protected ResponseEntity<String> put(String url) {
        return restTemplate.exchange(url, HttpMethod.PUT, HttpEntity.EMPTY, String.class);
    }

    protected void givenAnExistingBookWith(String id, String title) {
        post("{'id': " + id + ", 'title': " + title + "}");
        weWaitForEventsToBeProcessed();
    }

    protected void weWaitForEventsToBeProcessed() {
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: src/test/java/com/letgo/book/acceptance/TestCase.kt
package com.letgo.book.acceptance

import com.letgo.book.domain.BookRepository
import com.letgo.book.unit.domain.ABook
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.web.server.LocalServerPort
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity

private const val DOMAIN_NAME = "http://localhost:"

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
internal abstract class TestCase {
    @LocalServerPort
    private var port = 0

    @Autowired
    private lateinit var restTemplate: TestRestTemplate

    @Autowired
    private lateinit var bookRepository: BookRepository

    protected fun get(endpoint: String): ResponseEntity<String> =
        restTemplate.getForEntity(apiEndpoint(endpoint), String::class.java)

    protected fun post(body: String): ResponseEntity<String> {
        val headers = HttpHeaders().also {
            it.contentType = MediaType.APPLICATION_JSON
        }
        return restTemplate.exchange(
            apiEndpoint("/book"),
            HttpMethod.POST,
            HttpEntity(body, headers),
            String::class.java
        )
    }

    protected fun put(endpoint: String): ResponseEntity<String> =
        restTemplate.exchange(endpoint, HttpMethod.PUT, HttpEntity.EMPTY, String::class.java)

    protected fun givenAnExistingBookWith(id: String, title: String) =
        ABook.with(id = id, title = title).also {
            bookRepository.save(it)
        }

    protected fun weWaitForMessagesToBeProcessed() = Thread.sleep(1000)

    private fun apiEndpoint(endpoint: String) = DOMAIN_NAME + port + endpoint
}
